File: 99-recover-binary-search-tree/99-recover-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
  public void recoverTree(TreeNode root) {  

    TreeNode first = null;     // first node need to be swap
    TreeNode second = null;    // second node need to be swap
    TreeNode pre = new TreeNode(Integer.MIN_VALUE);  //previous node.

    while(root!=null){
      TreeNode node = root.left;

      // If left is not null, we need to find the rightmost node of left subtree,
      // Set its right child to current node
      if(node!=null){ 
     
        //find the rightmost
        while(node.right!=null && node.right != root){
          node = node.right;
        }

        //There are two cases,
        //null: first time we access current, set node.right to current and move to left child of current and continue;
        //current: we accessed current before, thus we've finished traversing left subtree, set node.right back to null;
        if(node.right == null){
          node.right = root;
          root = root.left;
          continue;
        }else{
          node.right = null;
        }
      }

      // compare current node with previous node
      if(root.val < pre.val ){
        // first time we enconter reversed order, we set previous node to first
        if( first == null ){
          first = pre;
        }
        //In case that two nodes are successive, we set second to current every time.
        second = root;
      }
      pre = root;
      root = root.right;      
    }

    //swap the value;
    int temp = second.val;
    second.val = first.val;
    first.val = temp;
  }
}

File: 99-recover-binary-search-tree/99-recover-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* first;
    TreeNode* middle;
    TreeNode* last;
    TreeNode* prev;
    void inorder(TreeNode *head)
    {
        if(head==NULL) return;
        
        inorder(head->left);
        if(prev!=NULL && head->val< prev->val)
        {
            // non adjacent case
            
            if(first==NULL)
            {
                first=prev;
                middle=head;
                
            }
        
            // adjacent case
            else
            {
                last=head;
            }
            
        }
        prev=head;
        inorder(head->right);
    }
    
    void recoverTree(TreeNode* root)
    {
        first=middle=last=NULL;
        prev=new TreeNode(INT_MIN);
        inorder(root);
        // NOT adjacent case
        
        if(first!=NULL && last!=NULL) swap(first->val,last->val);
        else if(first!=NULL && middle!=NULL) swap(first->val,middle->val);
        
    }
};